sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala

```diff
@@ -0,0 +1,145 @@
+  // that we have already processed).
+  {
+    val storedProgress = sink.currentProgress
+    sources.foreach { s =>
+      storedProgress.get(s).foreach { offset =>
+        currentOffsets.update(s, offset)
+      }
+    }
+  }
+
+  logInfo(s"Stream running at $currentOffsets")
+
+  /** When false, signals to the microBatchThread that it should stop running. */
+  @volatile private var shouldRun = true
+
+  /** The thread that runs the micro-batches of this stream. */
+  private[sql] val microBatchThread = new Thread("stream execution thread") {
+    override def run(): Unit = {
+      SQLContext.setActive(sqlContext)
+      while (shouldRun) { attemptBatch() }
+    }
+  }
+  microBatchThread.setDaemon(true)
+  microBatchThread.start()
+
+  var lastExecution: QueryExecution = null
+
+  /**
+   * Checks to see if any new data is present in any of the sources.  When new data is available,
+   * a batch is executed and passed to the sink, updating the currentOffsets.
+   */
+  private def attemptBatch(): Unit = {
+
+    val newData = sources.flatMap { s =>
+      val prevOffset = currentOffsets.get(s)
+      val latestOffset = s.getCurrentOffset
+      if (prevOffset.isEmpty || latestOffset > prevOffset.get) {
+        Some(s -> latestOffset)
+      } else None
+    }.toMap
+
+    if (newData.nonEmpty) {
+      val startTime = System.nanoTime()
+      logInfo(s"Running with new data up to: $newData")
+
+      // Replace sources in the logical plan with data that has arrived since the last batch.
+      val newPlan = logicalPlan transform {
+        case StreamingRelation(source, output) =>
+          if (newData.contains(source)) {
+            val batchInput =
+              source.getSlice(sqlContext, currentOffsets.get(source), newData(source))
+            LogicalRDD(output, batchInput)(sqlContext)
+          } else {
+            LocalRelation(output)
+          }
+      }
+
+      val optimizerStart = System.nanoTime()
+
+      lastExecution = new QueryExecution(sqlContext, newPlan)
+      val executedPlan = lastExecution.executedPlan
+      val optimizerTime = (System.nanoTime() - optimizerStart).toDouble / 1000000
+      logDebug(s"Optimized batch in ${optimizerTime}ms")
+
+      val results = executedPlan.execute().map(_.copy())
+      newData.foreach(currentOffsets.update)
+      sink.addBatch(currentOffsets.copy(), results)
+
+      StreamExecution.this.synchronized {
+        StreamExecution.this.notifyAll()
+      }
+
+      val batchTime = (System.nanoTime() - startTime).toDouble / 1000000
+      logInfo(s"Compete up to $newData in ${batchTime}ms")
+    }
+
+    logDebug(s"Waiting for data, current: $currentOffsets")
+    Thread.sleep(10)
+  }
+
+  /**
+   * Signals to the thread executing micro-batches that it should stop running after the next
+   * batch. This method blocks until the thread stops running.
+   */
+  def stop(): Unit = {
+    shouldRun = false
+    while (microBatchThread.isAlive) { Thread.sleep(100) }
```

Comment:
`sleep` can be replaced with `join`: `if (microBatchThread.isAlive) microBatchThread.join()`